src/WebTelemetryBase.ts

```diff
@@ -95,36 +94,51 @@ export abstract class WebTelemetryBase<P, R> {
         }
     }
 
-    protected createEvent<M>(payload: P, meta?: M): WebTelemetryBaseEvent {
+    protected createEvent<M>(payload: P, meta?: M): Promise<WebTelemetryBaseEvent> {
         let evt: WebTelemetryBaseEvent = {
             sessionId: globalSessionId,
             ...this.payloadToJSON(payload),
         };
 
         let evtMetadata = meta || {};
 
-        for (const addon of this.addons) {
-            const data = addon.data();
-            const metadata = addon.metadata();
-            evt = Object.assign({}, data, evt);
-            evtMetadata = Object.assign({}, metadata, evtMetadata);
-        }
-
-        evt.metadata = stringifyCircularObj(evtMetadata);
-
-        return evt;
+        const addonPromises = this.addons.map((addon) => Promise.all([addon.data(), addon.metadata()]));
+
+        return Promise.all(addonPromises).then((results) => {
+            const [finalEvt, finalMetadata] = results.reduce(
+                ([currentEvt, currentMetadata], [dataResult, metadataResult]) => {
+                    return [
+                        {
+                            ...dataResult,
+                            ...currentEvt,
+                        },
+                        {
+                            ...metadataResult,
+                            ...currentMetadata,
+                        },
+                    ];
+                },
+                [{}, evtMetadata],
+            );
+
+            return {
+                ...{ ...evt, ...finalEvt },
+                metadata: stringifyCircularObj(finalMetadata),
+            };
+        });
     }
```

Comment:
Логика с `counter` уж какая-то сильно мудрёная. Может положиться на `Promise.all`? И ещё не вижу смысла использовать `Object.assign`

```suggestion
    protected sendHandler() {
        if (!this.addons.length) {
            this.callTransport({});
            return;
        }
    
        Promise.all(
            this.addons.map(addon => Promise.all([addon.data(), addon.metadata()]))
        ).then(results => {
            results.forEach(([dataResult, metadataResult]) => {
                this.mainEvent = { ...dataResult, ...this.mainEvent };
                this.metaData = { ...metadataResult, ...this.metaData };
            });
    
            const event = {
                ...this.mainEvent,
                metadata: stringifyCircularObj(this.metaData),
            };
    
            this.callTransport(event);
        }).catch(error => {
            console.error('Error in sendHandler:', error);
            this.callTransport({
                ...this.mainEvent,
                error: 'Failed to process addons',
                metadata: stringifyCircularObj(this.metaData),
            });
        });
    }
```

Reply:
@arsenykruglikov Арсений. Это ты про какое место? 
 > Логика с counter уж какая-то сильно мудрёная

Reply:
@SeanSilke Он не там комментарий наверно оставил, он имел ввиду WebTelemetryApp.ts

Reply:
У нас там логика с counter